### DISCUSSION TITLE
tutorials/lift-off-part1/04-building-our-schema

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#30] [course feedback]
# tutorials/lift-off-part1/04-building-our-schema

https://www.apollographql.com/tutorials/lift-off-part1/04-building-our-schema

**Comment by Jma529** [comment]
Hi there, I've installed gql and Apollo server, but I can't seem to get the syntax highlighting to work, any ideas?

**Reply by leo-nestor** [reply]
GraphQL: Syxtax Highlighting extension works for me

**Reply by nichmorgan** [reply]
Download or search in vscode the extension below:

Nome: Apollo GraphQL
ID: apollographql.vscode-apollo
Descrição: Rich editor support for GraphQL client and server development that seamlessly integrates with the Apollo platform
Versão: 1.19.11
Editor: Apollo GraphQL
Link do Marketplace do VS: https://marketplace.visualstudio.com/items?itemName=apollographql.vscode-apollo

**Reply by yunushabeeb** [reply]
Thanks for the extension. @nichmorgan

**Reply by baizhi958216** [reply]
Thanks! @nichmorgan, There are lots of GraphQL extensions, but use apollographql can get syntax highlight

**Reply by soumayabar-rhoute** [reply]
thank you for the extension

**Comment by attilacsanyi** [comment]
Nice section! Maybe worth mention that `spaceCats` **can be null** when requesting to write the schema manually as it is easy to assume should not be empty based on the Track/Author example.

**Comment by Minifish57** [comment]
Hey, nice course and very useful information, i loved all that i learned here.
One question remains, not related to graphql and apollo, but what is the tool you use to write , draw, and present the information the funny way ?

**Comment by adriatic** [comment]
Any expression like
```
"A track is a group of Modules that teaches about a specific topic"
type Track {
  # Fields go here
}
```

results with error `Type aliases can only be used in TypeScript files.ts`

This seems to be an old VSCode problem and I wonder what is your solution to avoid it.

**Reply by mabuyo** [reply]
Hey @adriatic ! This error pops up when the SDL is _outside_ of the `gql` tag. To fix it, make sure the code snippet you included is within the `gql` tag. In `schema.js`, it should look something like this:

```js
const typeDefs = gql`
  "A track is a group of Modules that teaches about a specific topic"
  type Track {
    # Fields go here
  }
`;

module.exports = typeDefs;
```

Hope that helps!

**Reply by adriatic** [reply]
Hallo, Michelle

Thanks for your tip. If I choose to use it, any schema definition file would have to be changed, so I wonder what was this not already done in the original text. 

Regards


I believe that it setting the VSCode according to [this post](https://github.com/Microsoft/vscode-react-native/issues/631#issuecomment-372342868) is a cleaner approach - as it is problem by Microsoft (and VSCode team).

**Reply by mabuyo** [reply]
I'm not sure what you mean @adriatic. The error should not appear if the SDL code is within the `gql` tag. The SDL _should_ be within the `gql` tag, otherwise, the code won't work.

The other alternative is to write your SDL in a `.graphql` file. In that case, we would not need the `gql` tag and the error would not appear.

**Comment by adriatic** [comment]
Hi, Michelle

Please check the complete definition of this schema in your tutorial, and observe that is contained in the .js file and that the code is not inside `gql` tag.

**Reply by mabuyo** [reply]
Sorry for the confusion! I'll update the course shortly for clarity, as it's a bit hidden in the instructions where the schema is supposed to go.

```
const typeDefs = gql`
  # Schema definitions go here
`;
module.exports = typeDefs;
```

**Comment by wesley974** [comment]
Hi,

One thing that I don't understand well, why sometimes we use in Query a list named [Something!]! and sometimes named [Anything] ?
(I mean with the both '!' and sometime without both '!')

**Reply by repiemm** [reply]
It depends on the business requirements, for example, if your server expects that a List of Something will be always returned, then you need to specify that on the schema by adding the "!" character in front of both the content list and the list, but it really depends on what you need to be returned. More info about this here: https://graphql.org/learn/schema/#lists-and-non-null

**Reply by belloshehu** [reply]
Having two exclamation marks, one inside and another outside, means that the query must not return an empty list or a list containing null values. It simply means a list with non-null value must be returned.

**Reply by lizhennessy** [reply]
In this case: `[Track!]!`,  an empty array _is_ a possible and valid return value. The array _itself_ however cannot be null.

```
// Valid return values
[]
[{ id: "track-1", title: "Catstronauts", thumbnail: "http://..." }]

// invalid return values
null
[null]
```

**Reply by belloshehu** [reply]
Thanks for the correction and clarification! Pardon my misplacement of terms.  I often found my self using the term "list" in place of "array" because of my past python experience.

**Reply by lizhennessy** [reply]
No misuse of terms at all @belloshehu ! 😊 You're correct in that `[]` in the `[Track]` syntax is considered the List type! Thanks for your contributions to our learner conversations. ⭐

**Reply by belloshehu** [reply]
I am glad that I did not misuse terms. It is my pleasure to participate, contribute and share my view with others.

**Comment by saiyantanghosh** [comment]
done

**Comment by marksikaundi** [comment]
I missed where "define the Mission type with its subfields: id of type ID!, name of type String!, and description of type String!." But i believe with time am getting to understand

**Reply by repiemm** [reply]
Were you able to solve it?

**Comment by EjDadivas** [comment]
Summary of what I learned:
1. **gql** from "graphql-tag" - tagged template literal, used for wrapping GraphQL String -> converts it to format for Apollo Libraries
2. When Organizing data into types:
   - Think in terms of *standalone entities*
   - Schemas should reflect *business domain rules*
3. **type Query** - special type that tells GraphQL server what to retrieve.
   - fields here are the *entry points* that the client can query

**Comment by asingk** [comment]
so what is the difference between schema.ts and schema.graphql?

**Reply by mabuyo** [reply]
Hey there @asingk. We're only working with the `schema.ts` file in the course. However, you might see patterns where the schema is defined in a `.graphql` file. This means we don't need to use the `gql` template literal tag to enclose our schema. Instead, we would import the `schema.graphql` file in and connect it with our server initialization. 

You can find an example of this in our Voyage I course, where the [schema is defined here](https://github.com/apollographql/odyssey-voyage-I/blob/main/subgraph-locations/locations.graphql), and [loaded in here](https://github.com/apollographql/odyssey-voyage-I/blob/main/subgraph-locations/index.js#L7). Hope that helps!

**Comment by lycha0206** [comment]
is it true that over here, you can use `import graphql-tag` and `export default` instead of `require()` and `module.exports`? Can you put a sidenote here as to how it can be done?

**Reply by lizhennessy** [reply]
Hi there @lycha0206! In this case, whether we use `import`/`export` syntax or `require`/`module.exports` depends on how we're resolving files in our project. The JavaScript version of the course uses CommonJS (i.e. the `require` syntax), whereas TypeScript uses modules. Please feel free to switch to the TS version if you'd like to see how this code is put together!

**Comment by NishantCoder108** [comment]
# Building Our Schema 

## Important points I learned
 - `graphql-tag`  is used for as a template , where we write all types or query

 - ``` Query write for client side from which ,they will fetch specific they are written before in backend
      type Query {
      trackOfHome : [Track]
   }

- write all types  ,that are necessary. eg.
 ``` 
   type Author {
       id: ID!
      name: String!
     img: String
 }

 -

**Comment by pvnotpv** [comment]
oh dear finally found the right guide

**Comment by fredo-** [comment]
Might want to add a note about installing extension in IDE to enable SDL highlighting!

**Reply by fredo-** [reply]
To the video/lesson body

**Comment by jorge-rojas-TWM** [comment]
About the question, Which of these are always true about the Query type?
Its fields **cannot actually be null**:
  type Query {
    "Get tracks array for homepage grid"
    tracksForHome: [Track!]!
  }
tracksForHome is its only field. It may be empty, but not null (!)

**Reply by lizhennessy** [reply]
Hi there @jorge-rojas-TWM ! I see how this question is a little confusing. It's intent is to ask what is _always_ true about the `Query` type; not just the `Query` type that we've built in this lesson. So in this case, we might reword the incorrect choice ("Its fields cannot be null") to ("Its fields cannot return null"). We do need to give every Query type at least one field, however it's perfectly valid for those fields to _return_ `null`. I'll bring this back to the team so that we can make the update! Thanks for bringing this to our attention.

**Comment by virkumpavat** [comment]
Great , So far going well
